### DISCUSSION TITLE
std::span in default bindings?

### DISCUSSION THREAD

**Discussion by NAThompson** [pybind/pybind11#4177] [Ideas]
Currently, the [list of all builtin conversions](https://pybind11.readthedocs.io/en/stable/advanced/cast/overview.html#conversion-table) has a large number of extremely useful converters, however `std::span` has not yet been implemented.

Would it be sensible to add `std::span`? It appears to bind more naturally to numpy vectors than does `std::vector` and might be "the right way" to expose C++ memory to python in the future.

**Comment by Skylion007** [comment]
Would it be possible to have the std::span integration contained in its own header? If so, we could easily add it like we do for std::filesystem.

**Reply by NAThompson** [reply]
That would be more than acceptable for my use case; and I imagine would would well for everyone else.

**Reply by Skylion007** [reply]
We would gladly welcome a PR then.

**Comment by virtuald** [comment]
We use an LLVM backport of std::span, including conversions to/from buffer protocol objects. I don't want to make the PR myself, but happy for someone else to do it.

* Type caster: https://github.com/robotpy/robotpy-wpiutil/blob/main/wpiutil/src/type_casters/wpi_span_type_caster.h
* C++ portion of tests: https://github.com/robotpy/robotpy-wpiutil/blob/main/tests/cpp/wpiutil_test/module.cpp#L24
* Python portion of tests: https://github.com/robotpy/robotpy-wpiutil/blob/main/tests/test_span.py

**Reply by virtuald** [reply]
I did upgrade mine to std::span recently, still the same filenames.

Also, pybind11-abseil has a span implementation -- which also has a zero-copy conversion to/from buffer. Good reading material: https://github.com/pybind/pybind11_abseil/blob/master/pybind11_abseil/absl_casters.h#L317
